### PR TITLE
test: e2e subdomain redirect tests

### PR DESCRIPTION
**Pull request by SgtPooki**
- test: add reverse-proxy for e2e tests
- test: subdomain-detection test
- fix: do not render redirectPage if not parent page
- test: subdomain-detection with autoreload test

## Title

test: e2e subdomain redirect tests

## Description

subdomain redirect tests related to https://github.com/ipfs-shipyard/helia-service-worker-gateway/issues/134

### Summary of changes

1. add reverse-proxy for e2e tests
2. add subdomain-detection test
3. do not render redirectPage if not parent page
4. add subdomain-detection with autoreload test
5. move locators for e2e tests to a separate file

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: playwright.config.js [hunk 1]
```diff
@@ -18,13 +18,15 @@ export default defineConfig({
...
+    baseURL: 'http://localhost:3333',
 
     /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
     trace: 'on-first-retry',
 
     // 'allow' serviceWorkers is the default, but we want to be explicit
-    serviceWorkers: 'allow'
+    serviceWorkers: 'allow',
+
+    ignoreHTTPSErrors: true
```
Comment (SgtPooki) [inline playwright.config.js:29]: we don't actually need this anymore
```suggestion
```

File: src/sw.ts [hunk 2]
```diff
@@ -184,6 +184,11 @@ async function requestRouting (event: FetchEvent, url: URL): Promise<boolean> {
   } else if (!isValidRequestForSW(event)) {
     trace('helia-sw: not a valid request for helia-sw, ignoring ', event.request.url)
     return false
+  } else if (url.href.includes('bafkqaaa.ipfs')) {
+    // TODO: we shouldn't need to do this, but sometimes verified-fetch will attempt to fetch this.
```
Comment (SgtPooki) [inline src/sw.ts]: i was getting failures on the bafkqaaa "no gateways could find this block" but verified that helia and unixfs handles it properly (https://codepen.io/SgtPooki/pen/xxedpGP).

We need to investigate with verified-fetch.

File: src/sw.ts [hunk 3]
```diff
@@ -460,7 +465,7 @@ async function fetchHandler ({ path, request, event }: FetchHandlerArg): Promise
     const response = await verifiedFetch(verifiedFetchUrl, {
       signal,
       headers,
-      // TODO redirect: 'manual', // enable when http urls are supported by verified-fetch: https://github.com/ipfs-shipyard/helia-service-worker-gateway/issues/62#issuecomment-1977661456
+      redirect: 'manual',
```
Comment (SgtPooki) [inline src/sw.ts:471]: finally done with this todo

File: test-e2e/fixtures/set-sw-config.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          resolve()
+          channel.removeEventListener('message', onSuccess)
+        }
+      }
+      channel.addEventListener('message', onSuccess)
+    })
+    channel.postMessage({ target: 'SW', action: 'RELOAD_CONFIG', source: 'WINDOW' })
+    await swResponsePromise
+    // TODO: we shouldn't need this. We should be able to just post a message to the service worker to reload it's config.
+    window.postMessage({ source: 'helia-sw-config-iframe', target: 'PARENT', action: 'RELOAD_CONFIG', config: configInPage }, { targetOrigin: window.location.origin })
```
Comment (SgtPooki) [inline test-e2e/fixtures/set-sw-config.ts:62]: this is actually the only part that ~~works~~ triggers the config reload on subdomains.. without this, it doesn't work in the autoreload test, however, the above is needed to set the config on origin domain.

I confirmed that origin is getting indexedDB set properly, but autoreload doesn't get picked up by subdomain for some reason.

File: test-e2e/path-routing.test.ts [hunk 5]
```diff
@@ -3,10 +3,10 @@ import { waitForServiceWorker } from './fixtures/wait-for-service-worker.js'
 
 test.describe('path-routing', () => {
   test('can load identity CID via path', async ({ page }) => {
-    await page.goto('http://127.0.0.1:3000', { waitUntil: 'networkidle' })
+    await page.goto('http://127.0.0.1:3333', { waitUntil: 'networkidle' })
     // wait for service worker to load
     await waitForServiceWorker(page)
-    const response = await page.goto('http://127.0.0.1:3000/ipfs/bafkqablimvwgy3y', { waitUntil: 'networkidle' })
+    const response = await page.goto('http://127.0.0.1:3333/ipfs/bafkqablimvwgy3y', { waitUntil: 'networkidle' })
```
Comment (SgtPooki) [inline test-e2e/path-routing.test.ts:10]:
```suggestion
    const response = await page.goto('/ipfs/bafkqablimvwgy3y', { waitUntil: 'networkidle' })
```

File: test-e2e/path-routing.test.ts [hunk 6]
```diff
@@ -3,10 +3,10 @@ import { waitForServiceWorker } from './fixtures/wait-for-service-worker.js'
 
 test.describe('path-routing', () => {
   test('can load identity CID via path', async ({ page }) => {
-    await page.goto('http://127.0.0.1:3000', { waitUntil: 'networkidle' })
+    await page.goto('http://127.0.0.1:3333', { waitUntil: 'networkidle' })
```
Comment (SgtPooki) [inline test-e2e/path-routing.test.ts:7]:
```suggestion
    await page.goto('/', { waitUntil: 'networkidle' })
```

File: test-e2e/reverse-proxy.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
+/**
+ * This is a simple reverse proxy that makes sure that any request to localhost:3333 is forwarded to localhost:3000.
```
Comment (SgtPooki) [inline test-e2e/reverse-proxy.js:2]: this should also make it easier for devs to contribute without having to install traefik/nginx/caddy or do anything fancy

File: test-e2e/subdomain-detection.test.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    await page.goto('/', { waitUntil: 'networkidle' })
+    await waitForServiceWorker(page)
+
+    const initialResponse = await page.goto('/ipfs/bafkqablimvwgy3y', { waitUntil: 'commit' })
+
+    expect(initialResponse?.url()).toBe('http://bafkqablimvwgy3y.ipfs.localhost:3333/')
+    expect(initialResponse?.request()?.redirectedFrom()?.url()).toBe('http://localhost:3333/ipfs/bafkqablimvwgy3y')
+
+    await page.waitForURL('http://bafkqablimvwgy3y.ipfs.localhost:3333')
+    await setConfig({ page, config: { autoReload: true } })
```
Comment (SgtPooki) [inline test-e2e/subdomain-detection.test.ts:38]: ideally, `setConfig` would occur immediately after line 30, but.. alas, subdomains in playwright are not picking up the config for some reason.

File: test-e2e/fixtures/set-sw-config.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    const swResponsePromise = new Promise<void>((resolve, reject) => {
+      const onSuccess = (e: MessageEvent): void => {
+        if (e.data.action === 'RELOAD_CONFIG_SUCCESS') {
+          resolve()
+          channel.removeEventListener('message', onSuccess)
+        }
+      }
+      channel.addEventListener('message', onSuccess)
+    })
+    channel.postMessage({ target: 'SW', action: 'RELOAD_CONFIG', source: 'WINDOW' })
```
Comment (SgtPooki) [inline test-e2e/fixtures/set-sw-config.ts:59]: I hope this code here helps people understand the benefit of `HeliaServiceWorkerCommsChannel`, even if it does need cleaned up further.

**Review by SgtPooki — COMMENTED** [review]
self review

File: playwright.config.js [hunk 10]
```diff
@@ -38,12 +40,22 @@ export default defineConfig({
...
-    // need to use built assets due to service worker loading issue.
-    // TODO: figure out how to get things working with npm run start
-    command: 'npm run build && npx http-server --silent -p 3000 dist',
-    port: 3000,
-    timeout: 120 * 1000,
-    reuseExistingServer: !process.env.CI
-  }
+  webServer: [
+    {
+      command: 'node test-e2e/reverse-proxy.js',
```
Comment (2color) [inline playwright.config.js:45]: Why is the reverse proxy needed? Currently it only seems to add CORS headers and allow OPTIONS requests.

File: src/sw.ts (hunk 3 above)
Comment (2color) [reply src/sw.ts:471]: That would close https://github.com/ipfs-shipyard/helia-service-worker-gateway/issues/62

File: playwright.config.js (hunk 10 above)
Comment (SgtPooki) [reply playwright.config.js:45]: without the reverse proxy, the subdomain origin isolation check will fail:

https://github.com/ipfs-shipyard/helia-service-worker-gateway/blob/89b3ace21ba51a705a0625d04df797c504d154f8/src/sw.ts#L402-L413

Comment (SgtPooki) [reply playwright.config.js:45]:
```bash
╰─ ✔ ❯ npm run test:chrome -- -g 'subdomain-detection' --max-failures 5

> helia-service-worker-gateway@1.0.0 test:chrome
> playwright test -c playwright.config.js --project chromium -g subdomain-detection --max-failures 5

[WebServer] (node:14358) [DEP0066] DeprecationWarning: OutgoingMessage.prototype._headers is deprecated
(Use `node --trace-deprecation ...` to show where the warning was created)

Running 2 tests using 1 worker

  ✘  1 [chromium] › subdomain-detection.test.ts:6:3 › subdomain-detection › path requests are redirected to subdomains (898ms)
  ✘  2 [chromium] › subdomain-detection.test.ts:28:3 › subdomain-detection › path requests are redirected to subdomains automatically with autoreload enabled (893ms)


  1) [chromium] › subdomain-detection.test.ts:6:3 › subdomain-detection › path requests are redirected to subdomains

    Error: expect(received).toBe(expected) // Object.is equality

    Expected: "http://bafkqablimvwgy3y.ipfs.localhost:3333/"
    Received: "http://localhost:3000/ipfs/bafkqablimvwgy3y"

      11 |     const initialResponse = await page.goto('/ipfs/bafkqablimvwgy3y', { waitUntil: 'commit' })
      12 |
    > 13 |     expect(initialResponse?.url()).toBe('http://bafkqablimvwgy3y.ipfs.localhost:3333/')
         |                                    ^
      14 |     expect(initialResponse?.request()?.redirectedFrom()?.url()).toBe('http://localhost:3333/ipfs/bafkqablimvwgy3y')
      15 |
      16 |     await page.waitForURL('http://bafkqablimvwgy3y.ipfs.localhost:3333')

        at /Users/sgtpooki/code/work/protocol.ai/ipfs-shipyard/helia-service-worker-gateway/test-e2e/subdomain-detection.test.ts:13:36

  2) [chromium] › subdomain-detection.test.ts:28:3 › subdomain-detection › path requests are redirected to subdomains automatically with autoreload enabled

    Error: expect(received).toBe(expected) // Object.is equality

    Expected: "http://bafkqablimvwgy3y.ipfs.localhost:3333/"
    Received: "http://localhost:3000/ipfs/bafkqablimvwgy3y"

      32 |     const initialResponse = await page.goto('/ipfs/bafkqablimvwgy3y', { waitUntil: 'commit' })
      33 |
    > 34 |     expect(initialResponse?.url()).toBe('http://bafkqablimvwgy3y.ipfs.localhost:3333/')
         |                                    ^
      35 |     expect(initialResponse?.request()?.redirectedFrom()?.url()).toBe('http://localhost:3333/ipfs/bafkqablimvwgy3y')
      36 |
      37 |     await page.waitForURL('http://bafkqablimvwgy3y.ipfs.localhost:3333')

        at /Users/sgtpooki/code/work/protocol.ai/ipfs-shipyard/helia-service-worker-gateway/test-e2e/subdomain-detection.test.ts:34:36

  2 failed
    [chromium] › subdomain-detection.test.ts:6:3 › subdomain-detection › path requests are redirected to subdomains
    [chromium] › subdomain-detection.test.ts:28:3 › subdomain-detection › path requests are redirected to subdomains automatically with autoreload enabled
```

Comment (SgtPooki) [reply playwright.config.js:45]: the ports are wrong, but you can see that it's not on a subdomain

File: src/sw.ts (hunk 3 above)
Comment (SgtPooki) [reply src/sw.ts:471]: i'll add an e2e test prior to closing #62

File: src/sw.ts (hunk 2 above)
Comment (SgtPooki) [reply src/sw.ts]: checked with verified fetch, and it seems to work fine there too, not sure why it fails in helia-service-worker-gateway, but this check is pretty lightweight and should prevent extra unnecessary network call: https://codepen.io/SgtPooki/pen/xxedpGP

will update comment

File: src/sw.ts (hunk 3 above)
Comment (SgtPooki) [reply src/sw.ts:471]: one last thing blocking fix of #62: https://github.com/ipfs/helia-verified-fetch/pull/33

it was previously only working for subpath directories.

**Pull request merged**